src/beast/beast/asio/streambuf.h

```diff
@@ -219,230 +264,417 @@ class basic_streambuf<Allocator>::element
+        return temp;
     }
 
-private:
-    friend class basic_streambuf;
-    mutable_buffers_type (basic_streambuf const& streambuf);
+    const_iterator&
+    operator--()
+    {
+        --it_;
+        return *this;
+    }
+
+    const_iterator
+    operator--(int)
+    {
+        auto temp = *this;
+        --(*this);
+        return temp;
+    }
 };
 
-template <class Allocator>
-basic_streambuf<Allocator>::mutable_buffers_type::mutable_buffers_type (
-    basic_streambuf const& streambuf)
-    : streambuf_ (&streambuf)
+template<class Allocator>
+basic_streambuf<Allocator>::mutable_buffers_type::mutable_buffers_type(
+    basic_streambuf const& sb)
+    : sb_(&sb)
+{
+}
+
+template<class Allocator>
+auto
+basic_streambuf<Allocator>::mutable_buffers_type::begin() const ->
+    const_iterator
 {
+    return const_iterator{*sb_, sb_->out_};
 }
 
-template <class Allocator>
+template<class Allocator>
 auto
-basic_streambuf<Allocator>::mutable_buffers_type::
-    transform::operator() (element const& e) const ->
-        value_type const
+basic_streambuf<Allocator>::mutable_buffers_type::end() const ->
+    const_iterator
 {
-    return value_type (e.data(), &e == &*std::prev(streambuf_->list_.end()) ?
-        streambuf_->out_end_ : e.size()) + (&e == &*streambuf_->out_ ?
-            streambuf_->out_pos_ : 0);
+    return const_iterator{*sb_, sb_->list_.end()};
 }
 
 //------------------------------------------------------------------------------
 
-template <class Allocator>
+template<class Allocator>
 basic_streambuf<Allocator>::~basic_streambuf()
 {
-    for(auto iter = list_.begin(); iter != list_.end();)
-    {
-        auto& e = *iter++;
-        size_type const n = e.alloc_size();
-        alloc_traits::destroy(this->member(), &e);
-        alloc_traits::deallocate(this->member(),
-            reinterpret_cast<char*>(&e), n);
-    }
+    delete_list();
 }
 
-template <class Allocator>
-basic_streambuf<Allocator>::basic_streambuf(std::size_t block_size,
-        Allocator const& alloc)
-    : empty_base_optimization<Allocator>(alloc)
-    , block_size_ (block_size)
-    , block_size_next_ (block_size)
-    , out_ (list_.end())
+template<class Allocator>
+basic_streambuf<Allocator>::basic_streambuf(basic_streambuf&& other)
+    : empty_base_optimization<Allocator>(other.member())
+    , alloc_size_(other.alloc_size_)
+    , in_size_(other.in_size_)
+    , in_pos_(other.in_pos_)
+    , out_pos_(other.out_pos_)
+    , out_end_(other.out_end_)
 {
-    if (! (block_size > 0))
-        throw std::invalid_argument(
-            "basic_streambuf: invalid block_size");
+    static_assert(std::is_copy_constructible<basic_streambuf>::value, "");
+    static_assert(std::is_move_constructible<basic_streambuf>::value, "");
+    static_assert(std::is_copy_assignable<basic_streambuf>::value, "");
+    static_assert(std::is_move_assignable<basic_streambuf>::value, "");
+
+    // VFALCO What about allocator propagate_on_container_move_construct?
```

Comment:
This is a fuzzy area.  If we want to go by the same rules that the std::containers go by, then no, the move assignment operator is not yet correct.  It should only move assign the allocator if `propagate_on_container_move_assignment` is true_type.  If it move assigns the allocator, then it can transfer memory ownership from the rhs.  I'm not positive what the move assignment of an intrusive list does, but I'm guessing it transfers memory ownership.

If `propagate_on_container_move_assignment` is false, then you don't move assign allocator.  But if the lhs and rhs allocators are equal, you can still transfer memory ownership (the list).  If the allocators are not equal (and `propagate_on_container_move_assignment` is false), no memory ownership transfer can take place.  You must reconstruct the value of the rhs on the lhs allocating from the lhs allocator as required to do so.